romancal/skymatch/tests/test_skymatch.py

```diff
@@ -0,0 +1,378 @@
+from itertools import product
+
+import astropy.units as u
+import numpy as np
+import pytest
+from astropy import coordinates as coord
+from astropy.modeling import models
+from gwcs import coordinate_frames as cf
+from gwcs import wcs as gwcs_wcs
+from roman_datamodels.datamodels import ImageModel, ModelContainer
+from roman_datamodels.maker_utils import mk_level2_image
+
+from romancal.lib import dqflags
+from romancal.skymatch import SkyMatchStep
+
+DO_NOT_USE = dqflags.pixel["DO_NOT_USE"]
+SATURATED = dqflags.pixel["SATURATED"]
+
+
+def mk_gwcs(shape, sky_offset=[0, 0] * u.arcsec, rotate=0 * u.deg):
+    # Example adapted from photutils:
+    #   https://github.com/astropy/photutils/blob/
+    #   2825356f1d876cacefb3a03d104a4c563065375f/photutils/datasets/make.py#L821
+    rho = np.pi / 3.0
+    # Roman plate scale:
+    scale = (0.11 * u.arcsec / u.pixel).to_value(u.deg / u.pixel)
+
+    shift_by_crpix = models.Shift((-shape[1] / 2) + 1) & models.Shift(
+        (-shape[0] / 2) + 1
+    )
+
+    cd_matrix = np.array(
+        [
+            [-scale * np.cos(rho), scale * np.sin(rho)],
+            [scale * np.sin(rho), scale * np.cos(rho)],
+        ]
+    )
+
+    rotation = models.AffineTransformation2D(cd_matrix, translation=[0, 0])
+    rotation.inverse = models.AffineTransformation2D(
+        np.linalg.inv(cd_matrix), translation=[0, 0]
+    )
+
+    tan = models.Pix2Sky_TAN()
+    celestial_rotation = models.RotateNative2Celestial(
+        197.8925 + sky_offset[0].to_value(u.deg),
+        -1.36555556 + sky_offset[1].to_value(u.deg),
+        180.0 + rotate.to_value(u.deg),
+    )
+
+    det2sky = shift_by_crpix | rotation | tan | celestial_rotation
+    det2sky.name = "linear_transform"
+
+    detector_frame = cf.Frame2D(
+        name="detector", axes_names=("x", "y"), unit=(u.pix, u.pix)
+    )
+
+    sky_frame = cf.CelestialFrame(
+        reference_frame=coord.ICRS(), name="icrs", unit=(u.deg, u.deg)
+    )
+
+    pipeline = [(detector_frame, det2sky), (sky_frame, None)]
+
+    return gwcs_wcs.WCS(pipeline)
+
+
+def mk_image_model(
+    rate_mean=0,
+    rate_std=1e-4,
+    sky_offset=[0, 0] * u.arcsec,
+    rotation=0 * u.deg,
+    image_shape=(100, 100),
+):
+    l2 = mk_level2_image(shape=image_shape)
+    l2_im = ImageModel(l2)
+    l2_im.data = u.Quantity(
+        np.random.normal(loc=rate_mean, scale=rate_std, size=l2_im.data.shape).astype(
+            np.float32
+        ),
+        l2_im.data.unit,
+    )
+
+    l2_im.meta["wcs"] = mk_gwcs(image_shape, sky_offset=sky_offset, rotate=rotation)
+
+    # fake a background until `rad` implements the schema:
+    l2_im.meta["background"] = dict(level=None, subtracted=False, method=None)
+
+    l2_im.meta.cal_step["skymatch"] = "INCOMPLETE"
+    return l2_im
+
+
+@pytest.fixture
+def wfi_rate():
```

Comment:
Why do a simple rename here?

Reply:
This might be an artifact of the tortured commit history here, happy to adjust as you see fit. 

I wanted to write the function `mk_image_model` which might be useful (or even moved) elsewhere, and a fixture `wfi_rate` for uniform use in any tests that may need the result of `mk_image_model`. 

Then I realized that in writing `mk_image_model`, I noticed most of the adjustments were adding metadata to the result of `mk_level2_image` (WCS, background, a `cal_step` update). So it occurred to me that there might one day be a `mk_level3_image` method, and then this function might not be necessary, and the fixture can call that function instead.

But even if there's a `mk_level3_image`, it will need a method like `mk_gwcs` in order to synthesize dithers with overlap for these tests. 

So in the end I decided not to prematurely optimize and leave these working methods in the tests.

Reply:
I am good with leaving it as is for future development.